compiler/src/iree/compiler/Codegen/Common/GPU/GPUMaterializeEncoding.cpp

```diff
@@ -340,6 +340,95 @@ struct GPUSetEncodingOpLoweringConversion
+
+  LogicalResult
+  matchAndRewrite(IREE::Encoding::UnsetEncodingOp unsetEncodingOp,
+                  OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto converter = static_cast<const MaterializeEncodingTypeConverter *>(
+        getTypeConverter());
+    auto unPackOp = lowerUnsetEncodingToUnpackOp(
+        rewriter, unsetEncodingOp, adaptor.getSource(), *converter,
+        this->materializeEncodingValueFn);
+    if (failed(unPackOp)) {
+      Value result = adaptor.getSource();
+      Type targetType =
+          getTypeConverter()->convertType(unsetEncodingOp.getResultType());
+      if (targetType != result.getType()) {
+        result = rewriter.create<tensor::CastOp>(unsetEncodingOp.getLoc(),
+                                                 targetType, result);
+      }
+      rewriter.replaceOp(unsetEncodingOp, result);
+      return success();
+    }
+
+    Location loc = unsetEncodingOp.getLoc();
+
+    FailureOr<MaterializeEncodingInfo> maybeEncodingInfo =
+        converter->getEncodingInfo(unsetEncodingOp.getSource().getType());
+    if (failed(maybeEncodingInfo)) {
+      return rewriter.notifyMatchFailure(unsetEncodingOp,
+                                         "unhandled result encoding");
+    }
+    SmallVector<int64_t> innerTiles = maybeEncodingInfo->innerTileSizes;
+    SmallVector<int64_t> intrinsicVectorShape =
+        maybeEncodingInfo->innerTileShapes;
+
+    rewriter.setInsertionPoint(*unPackOp);
+
+    auto sourceShape = unPackOp->getSourceType().getShape();
+    auto iT1 = intrinsicVectorShape[0];
+    auto iT2 = intrinsicVectorShape[1];
+    auto oT1 = sourceShape[2] / iT1;
+    auto oT2 = sourceShape[3] / iT2;
+    SmallVector<int64_t> transposeSourceDims = {
+        sourceShape[0], sourceShape[1], oT1, iT1, oT2, iT2};
+    assert(transposeSourceDims.size() == 6);
+
+    size_t targetRank = unsetEncodingOp.getResultType().getRank();
+
+    SmallVector<int64_t> transposePerm;
+    transposePerm.push_back(0);
+    transposePerm.push_back(1);
+    for (auto perm : maybeEncodingInfo->permutation) {
+      transposePerm.push_back(targetRank + perm);
+    }
+    SmallVector<int64_t> expandShapeResultDims = transposeSourceDims;
+    applyPermutationToVector(expandShapeResultDims, transposePerm);
+
+    auto expandShapeResultType = RankedTensorType::get(
+        expandShapeResultDims, unPackOp->getSourceType().getElementType());
+
+    tensor::ExpandShapeOp expandShapeOp =
+        reshapeForInnerDims(rewriter, unPackOp->getSource(),
+                            expandShapeResultType, unPackOp->getDestRank());
+
+    auto emptyTensor = rewriter.create<tensor::EmptyOp>(
+        loc, transposeSourceDims,
+        unsetEncodingOp.getSourceType().getElementType());
+    auto transposeOp = rewriter.create<linalg::TransposeOp>(
+        loc, expandShapeOp, emptyTensor,
+        invertPermutationVector(transposePerm));
```

Comment:
I'd move the inversion to l.401, which is more clear.

Reply:
done